x-pack/apm-server/aggregation/servicemetrics/aggregator.go

```diff
@@ -228,56 +247,55 @@ type metricsMapEntry struct {
 	aggregationKey
 }
 
-func (mb *metricsBuffer) storeOrUpdate(key aggregationKey, metrics serviceMetrics, logger *logp.Logger) bool {
+func (mb *metricsBuffer) storeOrUpdate(key aggregationKey, metrics serviceMetrics, logger *logp.Logger) {
 	// hash does not use the serviceMetrics so it is safe to call concurrently.
 	hash := key.hash()
 
 	// Full lock because serviceMetrics cannot be updated atomically.
 	mb.mu.Lock()
 	defer mb.mu.Unlock()
 
+	var entry *metricsMapEntry
 	entries, ok := mb.m[hash]
 	if ok {
 		for offset, old := range entries {
 			if old.aggregationKey.equal(key) {
-				entries[offset].serviceMetrics = serviceMetrics{
-					transactionDuration: old.transactionDuration + metrics.transactionDuration,
-					transactionCount:    old.transactionCount + metrics.transactionCount,
-					failureCount:        old.failureCount + metrics.failureCount,
-					successCount:        old.successCount + metrics.successCount,
-				}
-				return true
+				entry = entries[offset]
+				break
 			}
 		}
-	} else if mb.entries >= len(mb.space) {
-		return false
 	}
-
-	half := mb.maxSize / 2
-	if !ok {
-		switch mb.entries {
-		case mb.maxSize:
-			return false
-		case half - 1:
-			logger.Warn("service metrics groups reached 50% capacity")
-		case mb.maxSize - 1:
-			logger.Warn("service metrics groups reached 100% capacity")
+	if entry == nil && mb.entries >= mb.maxSize && mb.other != nil {
+		entry = mb.other
+		mb.otherCardinalityEstimator.Insert([]byte(key.serviceName))
+	}
+	if entry != nil {
+		entry.serviceMetrics = serviceMetrics{
+			transactionDuration: entry.transactionDuration + metrics.transactionDuration,
+			transactionCount:    entry.transactionCount + metrics.transactionCount,
+			failureCount:        entry.failureCount + metrics.failureCount,
+			successCount:        entry.successCount + metrics.successCount,
 		}
+		return
+	}
+	if mb.entries >= mb.maxSize {
+		mb.other = &mb.space[len(mb.space)-1]
+		mb.otherCardinalityEstimator = hyperloglog.New14()
+		mb.otherCardinalityEstimator.Insert([]byte(key.serviceName))
```

Comment:
> Should we be including all properties of the aggregation key in the cardinality?

I was also initially confused on this. I agree that capturing cardinality of service groups makes more sense. How do we plan to use this cardinality metrics?

Reply:
> How do we plan to use this cardinality metrics?

The idea is for the UI to query them and warn users when the limit is exceeded. The UI should be able to tell not just that the limit is exceeded, but also which service is producing the highest cardinality of transaction groups. That should help guide users in reconfiguring agents, fixing instrumentation, etc.

Ideally we would be able to aggregate cardinality over time, so we can tell if there's a constant but large number of services/transactions, vs. dynamic transaction names. That would require a new cardinality field type in Elasticsearch, i.e. enabling us to store HLL++ sketches, and then merge them at query time.

What we can still do though is compare the cardinality of shorter and longer time range metrics. If the longer time range metrics have higher cardinality, then that would imply there's some dynamicity in the transaction/service/whatever names.

Reply:
> That would require a new cardinality field type in Elasticsearch, i.e. enabling us to store HLL++ sketches, and then merge them at query time

_I think_ only this will also not work since the `other` bucket will not be deterministic for each aggregation interval given the same set of input if the number of services in the input exceeds the max services limit. Based on the order in which the services arrive different services will end up in `other` service name bucket.

> What we can still do though is compare the cardinality of shorter and longer time range metrics. If the longer time range metrics have higher cardinality, then that would imply there's some dynamicity in the transaction/service/whatever names.

Got it.

----

For the cardinality estimator, for service metrics I will update to use service group. What do you think about the transaction metrics? Will the `transaction.name` suffice for that?

Reply:
>> That would require a new cardinality field type in Elasticsearch, i.e. enabling us to store HLL++ sketches, and then merge them at query time
>
> I think only this will also not work since the other bucket will not be deterministic for each aggregation interval given the same set of input if the number of services in the input exceeds the max services limit. Based on the order in which the services arrive different services will end up in other service name bucket.

Ah, good point. Maybe we should instead be measuring cardinality of all service groups seen within the aggregation interval. The timestamp would be based on the server's timestamp, rather than the event timestamp. WDYT? (CC @felixbarny)

> For the cardinality estimator, for service metrics I will update to use service group. What do you think about the transaction metrics? Will the transaction.name suffice for that?

I think we should do the same thing for transactions, and use the full transaction group aggregation key.

Reply:
> I think we should do the same thing for transactions, and use the full transaction group aggregation key.

Updated the code for both txn and svc aggregation to use group aggregation key. One point to note is that now I am using `xxhash` (the hash created by us) rather than using the library's `metrohash`. This should make it a use a bit less CPU than before.

> Maybe we should instead be measuring cardinality of all service groups seen within the aggregation interval.

Hmm, I think recording `overflow` is better because if we were to emit full cardinality the users of the metrics need to be aware of our limit (or we will have to emit the limit as a separate metric). While possible, I am currently leaning towards emitting only overflow metrics. One additional point is that we do emit a metric from APM-server's instrumentation that tracks the overall overflow.

I also wanted to raise a concern about overflow into `other` service. This case indicates that there are too many services -> it is interesting because if it is genuine then we can't do much as we don't have an explicit config for this. Even with horizontal scaling, we can't fix this because all the services can go to all the scaled instances (this can be a case for having the service configuration as an explicit config (?))

Reply:
For the discussion around using event timestamp or processing timestamp, IIUC for the UI query to find how many metrics overflowed in, say, the last 30 minutes we will have to either use event processing timestamp or ingest timestamp.
I am thinking under the assumption that `ingest timestamp - processing timestamp` can be assumed to be a constant with higher probability than `processing timestamp - event timestamp`. So, we can visualize the data w.r.t processing timestamp by using the ingest timestamp however, we won’t have an alternative if we override the event timestamp with the processing timestamp.

Reply:
@lahsivjar I'm not sure I understand what you are proposing for the timestamp discussion.

Reply:
> The idea is for the UI to query them and warn users when the limit is exceeded. The UI should be able to tell not just that the limit is exceeded, but also which service is producing the highest cardinality of transaction groups. That should help guide users in reconfiguring agents, fixing instrumentation, etc.

@simitt I was thinking about how to answer the above two questions for the UI.  Both the questions, _I think_, will require the UI to use *APM Server's processing time* and will probably check data for the last xx duration of time. However, I was concerned that this would also affect the `other` bucket's histogram data since the metrics will be published in the same document as the `other` bucket's metrics.

After re-reading the comment and giving it a bit more thought I think we can publish the overflow metrics in a separate document with the timestamp as the server's processing time. WDYT? Let me know if I am still not clear.

Reply:
SGTM